src/Interpreters/Cache/QueryConditionCache.cpp

```diff
@@ -107,31 +99,72 @@ size_t QueryConditionCache::maxSizeInBytes()
     return cache.maxSizeInBytes();
 }
 
-bool QueryConditionCache::Key::operator==(const Key & other) const
-{
-    return table_id == other.table_id
-        && part_name == other.part_name
-        && condition_hash == other.condition_hash;
-}
+QueryConditionCacheWriter::QueryConditionCacheWriter(
+    QueryConditionCachePtr query_condition_cache_,
+    UInt64 condition_hash_,
+    const String & condition_)
+    : query_condition_cache(query_condition_cache_)
+    , condition_hash(condition_hash_)
+    , condition(condition_)
+{}
 
-QueryConditionCache::Entry::Entry(size_t mark_count)
-    : matching_marks(mark_count, true) /// by default, all marks potentially are potential matches, i.e. we can't skip them
+QueryConditionCacheWriter::~QueryConditionCacheWriter()
 {
-}
+    std::lock_guard<std::mutex> lock(mutex);
 
-size_t QueryConditionCache::KeyHasher::operator()(const Key & key) const
-{
-    SipHash hash;
-    hash.update(key.table_id);
-    hash.update(key.part_name);
-    hash.update(key.condition_hash);
-    return hash.get64();
+    for (const auto & [key, matching_marks] : buffer)
+    {
+        if (matching_marks.empty())
```

Comment:
This seems a strange case. I cannot imagine when the marks could be empty. Maybe we should instead assert that they are not.